_posts/2016-10-26-jwt-artsy-journey.md

```diff
@@ -0,0 +1,78 @@
+
+
+# Where we were
+
+After a successful login, our API would generate a custom JSON object including some basic information about the user and application, then encrypt it using a single server-side secret.
+For every authenticated request we would have to decrypt the token, make sure user and application are valid and application still has access to our API. We are already _stateless_ since we don't store tokens in our database.
+In this manner, we avoided needing to store sessions in the database. However, client applications had to make an API call to get any information about the authenticated user.
+
+
+# Where we are going
+We want to keep our current auth flow which is already stateless and mainly replace our in-house generated access token with a more standard JWT. What that gives us is:
+
+- Client applications are able to decode and use basic information out of our token. With our in-house generated access token this could happen only if client application had duplicated our API side's decryption logic. With JWT each app can decode the token and get basic data out of it, it can check if token is expired or not and etc. If client is doing something that needs to make sure token is still valid, it can still call the API, like before, to make sure token is valid and get more data about it.
+
+- Possibly include different data in JWT payload for different applications, if a JWT token was created for the application that deals with conversations, it doesn't need to know about the user's auction related information.
+
+- Follow a well-defined standard and use existing libraries for creating and reading the token. JWT has a decent set of reseverd _claims_ which can be used to describe the token in a unified language. Few examples:
+
+  - `exp`: Expiration time.
+  - `iat`: Time this JWT was issued.
+  - `iss`: Issuer of this JWT (ex. our main API)
+  - `aud`: Audience of this JWt (ex. Our mobile application)
+  - `jti`: JWT ID, unique identifier for this JWT
+
+Most JWT libraries honor these claims and can automatically validate them so we don't have to handle things like expiration ourselves.
+
+In our new approach, every `ClientApplication` has it's own `SecretKey`. When we get a new login for specific applicaiton:
+
+- We find the application in our database to make sure application is valid and still have access to our API.
+
+- We use this application's secret key and we generate a JWT
+
+On decoding and verification side, when we get an authenticated request:
+
+- We decode JWT without verifying the signature.
+
+- From JWT payload we get `aud` (audiance) attribute which defines for which application this JWT was generated.
+
+- We fetch `ClientApplication` for this `aud` and we verify JWT's signature using this application's `SecretKey` and make sure this is a valid and verified JWT.
```

Comment:
"...and make sure this is a valid..." isn't necessary, since you already say "verify JWT's signature."